rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,512 @@
+
+In order to correctly index the providers, the registry need to know what
+version of the Terraform Plugin API a given provider supports. This is not
+something we can deduce from the provider itself, and must be "just known"
+in advance.
+
+This will be injected into the build via a value in the Drone yaml. There
+is a danger that we will forget to modify this value should we upgrade the
+version of the Teleport Plugin framework we use, and mislabel a release in
+the registry.
+
+##### Signing GPG Key
+
+Packaging for the registry requires a GPG signing key, including an
+identity.
+
+Key is expected to be in PGP ASCII-armour format.
+
+The Identity is displayed to the user when the provider is installed, so it
+should be an "official" teleport key (or derived from one) rather than a
+self-signed key.
+
+For example (from my PoC):
+
+```shell
+Initializing provider plugins...
+- Finding terraform.clarke.mobi/gravitational/teleport versions matching "~> 8.3"...
+- Installing terraform.clarke.mobi/gravitational/teleport v8.3.4...
+- Installed terraform.clarke.mobi/gravitational/teleport v8.3.4 (self-signed, key ID 7DA6C64E1701F9E4)
+```
+
+#### Building the packages
+
+As far as the registry protocol is concerned, the entire package for a given
+provider consists of:
+
+* The actual binary distribution zipfile
+* A checksum file containing the SHA256 of the zipfile (formatted as per
+ `sha256sum` tool)
+* A signature file containing a binary, detached GPG signature for
+  checksum file.
+
+For each release tarball (one each for multiple release platforms - i.e.
+`linux` and `darwin`):
+
+1. deduce the version, OS and architecture of the release from the
+    release tarball filename
+2. repack the tarball into an appropriately formatted & named zipfile
+3. generate the checksum file from the zipfile
+4. generate the signature file by signing the checksum file
+
+#### Updating the registry
+
+To update the registry, we take the artefacts and data generated in the
+previous step and
+
+1. Download the existing `versions` index file (or create fresh, if
+    not present)
+2. Add entries for the new providers in to the index
+3. Create the `download` entries for each file
+
+I expect this process to be implemented a sa small Go program (or a small suite
+of them), rather than a shell script. Using small Go programs for build tasks
+has shown itself to be a useful, flexible and above all _legible_ tool for
+scripting build tasks in the mainline Teleport repository, and we should
+continue that pattern in `teleport-plugins`.
+
+#### Final publishing
+
+Upload the repackaged `zip` files, the updated `versions` index and the new
+`download` files to the S3 bucket.
+
+#### Synchronization Hazard
+
+The Download-Update-Publish cycle for updating the registry index is vulnerable
+to corruption if multiple independent updates occur simultaneously.
+
+Unfortunately, AWS S3 does not provide any built-in methods for preventing this
```

Comment:
I might be wrong, but as far as I remember, S3 has simple object locks with/without timeouts. For sure, they can't replace CAS, but might be helpful to fail fast and retry if an another conflicting build is in progress.

Reply:
I looked in to AWS locks, and I don't think they're suitable for our purposes. Will take a deeper look and see.